### PR TITLE
Introduce transfer gold page

### PR DESCRIPTION
**Pull request by riemannulus**
Prior to the release of the Ethereum Bridge, there was a demand for an interface that could transfer NCG from the launcher. This PR is the cornerstone for implementing that interface.

### CODE REVIEW

**Review by BasixKOR — COMMENTED** [review]
Just found some space for improvements.

File: src/transfer/hooks/index.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { MobXProviderContext } from "mobx-react";
+import TransferStore from "../stores/transfer";
+import MenuStore from "../stores/menu";
+
+interface IStoreContainer {
+    transferStore: TransferStore
+    menuStore: MenuStore
+}
+
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply src/transfer/hooks/index.ts:12]: Yes, It looks good. maybe we can refactor https://github.com/planetarium/9c-launcher/blob/development/src/hooks/useStores.ts here too.

File: tsconfig.json [hunk 2]
```diff
@@ -1,5 +1,9 @@
 {
   "compileOnSave": false,
+  "exclude": [
+    "node_modules",
+    "src/generated/**/*"
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply tsconfig.json:5]: Your reasoning is correct. It's for increase build speed.

File: src/transfer/stores/transfer.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    if (minerAddress.data) {
+      this.agentAddress = minerAddress.data.minerAddress;
+      return true;
+    }
+
+    return false;
+  };
+
+  @action
+  transferGold = async (recipient: string, amount: number, memo: string): Promise<boolean> => {
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply src/transfer/stores/transfer.ts:40]: Seems good

File: src/transfer/stores/transfer.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    this.checkAgentAddress();
+
+    const nextTxNonceData = await this.graphqlSdk.GetNextTxNonce({address: this.agentAddress});
+    if (!nextTxNonceData.data) {
+      throw new Error("Failed to get next nonce");
+    }
+    const txNonce = nextTxNonceData.data.transaction.nextTxNonce as string;
+    const tx = await this.graphqlSdk.Transfer({recipient, amount: amount.toString(), memo, txNonce});
+
+    return false;
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply src/transfer/stores/transfer.ts:50]: This is work-in-progress method.

File: src/transfer/stores/headless.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    if (!nextTxNonceData.data) {
+      throw new Error("Failed to get next nonce");
+    }
+    const txNonce = nextTxNonceData.data.transaction.nextTxNonce as string;
+    const tx = await this.graphqlSdk.Transfer({recipient, amount: amount.toString(), memo, txNonce});
+    if (!tx.data) {
+      throw new Error(`Failed to create transaction. ${tx.errors}`);
+    }
+
+    return tx.data.transfer;
```
Comment (BasixKOR) [inline src/transfer/stores/headless.ts:79]: I had to navigate few times to make sure what exactly this function is returning. How about adding a single type alias so it's more obvious what the function returns?

```ts
type TxId = string;

File: src/transfer/stores/mockHeadless.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import { observable } from "mobx";
+import { IHeadlessStore } from "./headless";
+
+export function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
```
Comment (BasixKOR) [inline src/transfer/stores/mockHeadless.ts:6]: You don't need to reimplement this; it's already in `utils.ts`.

**Review by BasixKOR — COMMENTED** [review]
Whoa that was a lot, I don't think anything needs some kimd of immidiate change, but I found some space for improvements.

And just a hyped note, but you could consider using XState for TransferPage management. It's already used in v2 Update phase management and I really recommend it :)

File: src/transfer/stores/transfer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (BasixKOR) [reply src/transfer/stores/transfer.ts]: The point still stands: I highly recommend using against number, Floating-point calculations may result in unintended amount of gold being transfered.

File: src/transfer/stores/mockHeadless.ts (hunk 6 above)
Comment (riemannulus) [reply src/transfer/stores/mockHeadless.ts:6]: Yes, it is implemented in `utils.ts`. but `utils.ts` have electron-depend import. So we cannot use `utils.ts` if we can want to use pure js. It is not intend, we need refactor about this.

File: src/transfer/stores/headless.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        case "INVALID":
+        case "STAGNING":
+          if(timeout) {
+            const elapsed = Date.now() - startTime;
+            if(elapsed >= timeout) {
+              onTimeout(txResult.blockIndex, txResult.blockHash as string);
+              return;
+            }
+          }
+          txResult = await (await this.graphqlSdk.TransactionResult({txId})).data!.transaction.transactionResult;
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply src/transfer/stores/headless.ts:120]: it is auto-generated code by GitHub Copilot. And I guess it is not intend. I'll change this method more readable.

File: src/transfer/stores/transfer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (riemannulus) [reply src/transfer/stores/transfer.ts]: Oh, I fogot about that. I'll change to decimal.js instead number.

**Comment by riemannulus** [comment]
> And just a hyped note, but you could consider using XState for TransferPage management. It's already used in v2 Update phase management and I really recommend it :)

Then can we apply XState first? IMO dependency conflict is hard to resolve.

**Comment by BasixKOR** [comment]
Yes, If you have a conflict in `yarn.lock` just run yarn once and yarn should resolve dependency conflicts.

**Comment by riemannulus** [comment]
PTAL 🙏  @BasixKOR

**Review by BasixKOR — APPROVED** [review]
Pretty awesome!

**Pull request merged**